consult-yasnippet.el

```diff
@@ -0,0 +1,158 @@
+          (yas-prompt-functions '(yas-no-prompt)))
+
+      ;; We always undo any snippet previews before maybe setting up
+      ;; some new previews.
+      (delete-region (car consult-yasnippet--region)
+                     (cdr consult-yasnippet--region))
+      (goto-char (car consult-yasnippet--region))
+      (setcar consult-yasnippet--region (point))
+      (insert consult-yasnippet--region-contents)
+      (setcdr consult-yasnippet--region (point))
+
+      (when template
+        (unwind-protect
+            (consult-yasnippet--expand-template template)
+          (unwind-protect
+              (mapc #'yas--commit-snippet
+                    (yas-active-snippets (point-min) (point-max)))
+            (setcdr consult-yasnippet--region (- (point-max) orig-offset))))
+        (redisplay)))))
+
+(defmacro consult-yasnippet--setup (&rest body)
+  "Setup the local variables and environment for `consult-yasnippet'.
+This environment is used both in `consult-yasnippet--preview'and
+`consult-yasnippet--expand-template'."
+  `(progn
+     (barf-if-buffer-read-only)
+     (unless (bound-and-true-p yas-minor-mode)
+       (error "`yas-minor-mode' not enabled in current buffer"))
+
+     (let* ((consult-yasnippet--snippets
+             (mapcar (lambda (template)
+                       (cons (yas--template-name template) template))
+                     (yas--all-templates (yas--get-snippet-tables))))
+            (consult-yasnippet--buffer (current-buffer))
+            (consult-yasnippet--region (if (region-active-p)
+                                           (cons (region-beginning) (region-end))
+                                         (cons (point) (point))))
+            (consult-yasnippet--region-contents (buffer-substring (car consult-yasnippet--region)
+                                                                  (cdr consult-yasnippet--region))))
+       ,@body)))
+
+(defun consult-yasnippet--read-snippet ()
+  "Backend implementation of `consult-yasnippet'.
+
+This starts a `completing-read' session with all the snippets in the current
+snippet table with support for previewing the snippet to be expanded and
+replacing the active region with the snippet expansion.
+
+This function doesn't actually expand the snippet, it only reads and then
+returns a snippet template from the user."
+  (consult-yasnippet--setup
+   (let ((buffer-undo-list t))                                                  ; Prevent querying user (and showing previews) from updating the undo-history
+     (unwind-protect
+         (consult--read
+          "Choose a snippet: "
+          consult-yasnippet--snippets
+          :lookup 'consult--lookup-cdr
+          :require-match t
+          :preview #'consult-yasnippet--preview
+          :category 'yasnippet)
+       (consult-yasnippet--preview nil t)))))                                   ; Restore contents of region from before preview (while still ignoring undo history).
```

Comment:
The :preview function has a bit of an unexpected feature - even if previews are disabled, the preview function is called with the restore=t flag in the end. Therefore you don't need the `(consult-yasnippet--preview nil t)` call.

Reply:
huh weird, while I was writing this I ran into issues with previews not being removed when you cancel completing-read. Guess that was because of something else. I'll remove this :-).

Reply:
@minad I still seem to have an issue with previews not being cleaned up after removing this. I'll push what I've got in a few minutes please double check whether there'es anything wrong with it.

Reply:
Regarding the arguments of the preview function, I should probably make this more clear:

* candidate=non-nil, restore=nil -> preview
* candidate=non-nil, restore=t -> restore and candidate has been selected
* candidate=nil, restore=t -> restore, abort, no candidate has been selected

Does this help?

Reply:
@minad 

That second case doesn't seem to be happening for me. Is `consult` passing the candidate in that case as well? That would explain the issue I'm having (`consult-yasnippet--preview` is then setting up the preview for that candidate which is then ignored by `consult-yasnippet` where it's re-entered). It's easier to demonstrate than explain.

Could you try:
1. Pulling this branch.
2. Requiring `consult-yasnippet`
3. Evaluate the following.
4. Try exanding a snippet.

```lisp
(defun consult-yasnippet--read-snippet ()
  (barf-if-buffer-read-only)

  (let* ((buffer-undo-list t))
    (consult--read
     "Choose a snippet: "
     (consult-yasnippet--candidates)
     :lookup 'consult--lookup-cdr
     :require-match t
     :preview (consult-yasnippet--preview)
     :category 'yasnippet)))
```

Reply:
I just checked. I added this to `consult-theme` and it seems to work as I described. Does this not work for you?

~~~ elisp
       :preview (lambda (cand restore)
                  (message "%S %S" cand restore)
~~~

I see the following messages:

~~~
some-theme nil
some-other-theme nil
....
selected-theme t OR nil t
~~~

Reply:
It works, but it also explains why we need to restore on exit.

```lisp
(consult--read
 "Foo: "
 '("a" "b" "c" "d")
 :preview (lambda (cand restore)
            (message "%S %S" cand restore)))
```

If you eval this your message buffer will look like:

```lisp
"b" nil
"a" nil
"c" nil
"d" nil
"c" nil
"a" nil
"a" t
"a"
```

At the end you pass the candidate and restore, but seeing as we restore before we preview a candidate we end up restoring, then previiweing, then when `consult-yasnippet` is called it expands the snippet in the middle of the preview.

The way things are implemented now we need to have that `unwind-protect` preview restore block.

Reply:
I think I understand now, I'll fix this :-).

Reply:
Fixed.

Reply:
> At the end you pass the candidate and restore, but seeing as we restore before we preview a candidate we end up restoring, then previiweing, then when consult-yasnippet is called it expands the snippet in the middle of the preview. The way things are implemented now we need to have that unwind-protect preview restore block.

I don't really understand what you mean. If restore=t is set and a candidate is passed, the operation should be committed. Otherwise it should just be reverted. But I think preview should be rewritten to use overlays, so you will end up with separate preview and final expansion code in the end. See `consult--yank-read` for the overlay preview code.

Reply:
Yeah I misunderstood the point of `restore`. I thought the order was something like:

```
1. restore
2. preview
3. restore
4. preview
5. ...
```

but it's more akin to you keep previewing until you restore. So you either restore or you preview, but I was doing both on each preview. That's a misunderstanding on my part, sorry. I've set it up to skip previewing when restore is true. Next should be overlays.